op-service/rethdb-reader/src/lib.rs

```diff
@@ -15,19 +16,76 @@ pub unsafe extern "C" fn rdb_read_receipts(
     block_hash: *const u8,
     block_hash_len: usize,
     db_path: *const c_char,
-) -> ReceiptsResult {
-    read_receipts_inner(block_hash, block_hash_len, db_path).unwrap_or(ReceiptsResult::fail())
+) -> *mut ReceiptsResult {
+    let res = read_receipts_inner(block_hash, block_hash_len, db_path);
+    match res {
+        Ok(res) => Box::into_raw(Box::new(res)),
+        Err(err) => Box::into_raw(Box::new(ReceiptsResult::fail(err.to_string()))),
+    }
 }
 
-/// Free a string that was allocated in Rust and passed to C.
+/// Free a ReceiptsResult that was allocated in Rust and passed to C.
 ///
 /// # Safety
 /// - All possible nil pointer dereferences are checked.
 #[no_mangle]
-pub unsafe extern "C" fn rdb_free_string(string: *mut c_char) {
-    // Convert the raw pointer back to a CString and let it go out of scope,
+pub unsafe extern "C" fn rdb_free_receipts(res: *mut ReceiptsResult) {
+    // Convert the raw pointer back to a ReceiptResult and let it go out of scope,
     // which will deallocate the memory.
-    if !string.is_null() {
-        let _ = std::ffi::CString::from_raw(string);
+    if !res.is_null() {
+        if !(*res).data.is_null() {
+            // Same deal with the data string
+            let _ = CString::from_raw((*res).data as *mut c_char);
+        }
+        if !(*res).error.is_null() {
+            // Same deal with the error string
+            let _ = CString::from_raw((*res).error);
+        }
+        let _ = Box::from_raw(res);
```

Comment:
Have we verified that the attempt to dereference the raw `ReceiptsResult` ptr after dropping its interior components? This seems like the ordering could be wrong.

Rust will drop both the `data` and `error` prior to the outer pointer, which could cause issues when attempting to reconstruct the `ReceiptsResult` afterwards. Not 100% sure at first glance, though.

```rs
if !res.is_null() {
    if !(*res).data.is_null() {
        // Same deal with the data string
        let _ = CString::from_raw((*res).data as *mut c_char);
        // <- above dropped
    }
    if !(*res).error.is_null() {
        // Same deal with the error string
        let _ = CString::from_raw((*res).error);
        // <- above dropped
    }
    // attempt to bring the `ReceiptsResult` back in, but it's interior values are already free'd!
    let _ = Box::from_raw(res);
    // <- above dropped
}
```

Reply:
I haven't verified this. But why would the `ReceiptsResult` be reconstructed after `data` and `error` are dropped?
This function consumes the ReceiptsResult pointer by freeing it. So it shouldn't be possible to reconstruct it elsewhere.

There are three objects created when users interface with this ffi:
- error
- data
- ReceiptsResult (containing the above)

Since both `error` and `data` live as unowned pointers, we need to manually free them prior to dropping the ReceiptsResult box.